transaction-view/src/sanitize.rs

```diff
@@ -0,0 +1,396 @@
+
+fn sanitize_instructions(view: &UnsanitizedTransactionView<impl TransactionData>) -> Result<()> {
+    // already verified there is at least one static account.
+    let max_program_id_index = view.num_static_account_keys().wrapping_sub(1);
+    // verified that there are no more than 256 accounts in `sanitize_account_access`
+    let max_account_index = total_number_of_accounts(view).wrapping_sub(1) as u8;
+
+    for instruction in view.instructions_iter() {
+        // Check that program indexes are static account keys.
+        if instruction.program_id_index > max_program_id_index {
+            return Err(TransactionViewError::SanitizeError);
+        }
+
+        // Check that the program index is not the fee-payer.
+        if instruction.program_id_index == 0 {
+            return Err(TransactionViewError::SanitizeError);
+        }
+
+        // Check that all account indexes are valid.
+        for account_index in instruction.accounts.iter().copied() {
+            if account_index > max_account_index {
+                return Err(TransactionViewError::SanitizeError);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn sanitize_address_table_lookups(
+    view: &UnsanitizedTransactionView<impl TransactionData>,
+) -> Result<()> {
+    for address_table_lookup in view.address_table_lookup_iter() {
+        // Check that there is at least one account lookup.
+        if address_table_lookup.writable_indexes.is_empty()
+            && address_table_lookup.readonly_indexes.is_empty()
+        {
+            return Err(TransactionViewError::SanitizeError);
+        }
+    }
+
+    Ok(())
+}
+
+fn total_number_of_accounts(view: &UnsanitizedTransactionView<impl TransactionData>) -> u16 {
+    u16::from(view.num_static_account_keys())
+        .saturating_add(view.total_writable_lookup_accounts())
+        .saturating_add(view.total_readonly_lookup_accounts())
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        crate::{
+            address_table_lookup_meta::AddressTableLookupMeta,
+            instructions_meta::InstructionsMeta,
+            message_header_meta::{MessageHeaderMeta, TransactionVersion},
+            signature_meta::SignatureMeta,
+            static_account_keys_meta::StaticAccountKeysMeta,
+            transaction_meta::TransactionMeta,
+        },
+        solana_sdk::{
+            hash::Hash,
+            message::{
+                v0::{self, MessageAddressTableLookup},
+                Message, MessageHeader, VersionedMessage,
+            },
+            pubkey::Pubkey,
+            signature::Signature,
+            system_instruction,
+            transaction::VersionedTransaction,
+        },
+    };
+
+    // Construct a dummy metadata object for testing.
+    // Most sanitization checks are based on the counts stored in this metadata
+    // rather than checking actual data. Using a mutable dummy object allows for
+    // far simpler testing.
+    fn dummy_metadata() -> TransactionMeta {
```

Comment:
Made several fields `pub(crate)` so that this style of testing is much simpler.
I think the above justifies that change, and public-ness is limited to this crate itself.